devdoc/channel.md

```diff
@@ -0,0 +1,294 @@
+# `channel` requirements
+
+## Overview
+
+`channel` is a module that allows the user to move data asynchronously from producer to consumer without having to block the thread.
+
+Consumers of data call `channel_pull` to register a `PULL_CALLBACK` function to be called when data becomes available.
+
+Producers of data call `channel_push` to notify the channel that data is available and provide a `PUSH_CALLBACK` to be called when the data has been consumed.
+
+`channel_pull` and `channel_push` can be called from different threads in any order. The callbacks are executed on threadpool threads.
+
+## `channel` internal state
+
+The internal state of the `channel` module can have the following values:
+  - `CHANNEL_EMPTY`: This is the initial state of the `channel`. It means that there is no data available to be pulled. Calling `channel_pull` or `channel_push` will cause the state to transition to `CHANNEL_PUSHED` or `CHANNEL_PULLED` respectively. After all the `channel_pull` and `channel_push` calls have been paired, the state will transition back to `CHANNEL_EMPTY`.
+  - `CHANNEL_PUSHED`: This is the state after `channel_push` has been called. A subsequent call to `channel_pull` will schedule the callbacks on the threadpool.
+  - `CHANNEL_PULLED`: This is the state after `channel_pull` has been called. A subsequent call to `channel_push` will the callbacks on the threadpool.
+
```

Comment:
Consider removing these states.

PULLED is equivalent with "data has been produced and is not consumed"

PUSHED is equivalent to "consumer of data exists but data is not produced yet".

Both of these imply existance of a previous call to _pull or to _push which implies "tail" of the queue not NULL.

tail!=NULL => if it has _pull callback=> PULLED state, otherwise it is PUSHED state.

Reply:
Removed